clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/IngestionBatchProcessor.java

```diff
@@ -71,32 +72,41 @@ public IngestionBatchProcessor(
     this.version = Version.parseVersionFromKafkaTopicName(storeVersionName);
   }
 
+  // For testing
+  KeyLevelLocksManager getLockManager() {
+    return this.lockManager;
+  }
+
   /**
    * When {@link #lockManager} is not null, this function will try to lock all the keys
    * (except Control Messages) passed by the params.
    */
-  public List<ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
+  public Map<ByteArrayKey, ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
     if (lockManager != null) {
-      List<ReentrantLock> locks = new ArrayList<>(records.size());
+      /**
+       * Need to use a {@link TreeMap} to make sure the locking will be executed in a deterministic order, otherwise
+       * deadlock can happen.
+       * Considering there could be multiple consumers, which are executing this function concurrently, and if they
+       * are trying to lock the same set of keys with different orders, deadlock can happen.
+       */
+      Map<ByteArrayKey, ReentrantLock> keyLockMap =
+          new TreeMap<>((o1, o2) -> ByteUtils.compare(o1.getContent(), o2.getContent()));
       records.forEach(r -> {
         if (!r.getKey().isControlMessage()) {
-          ReentrantLock lock = lockManager.acquireLockByKey(ByteArrayKey.wrap(r.getKey().getKey()));
-          locks.add(lock);
-          lock.lock();
+          keyLockMap.computeIfAbsent(ByteArrayKey.wrap(r.getKey().getKey()), k -> lockManager.acquireLockByKey(k));
         }
       });
-      return locks;
+      keyLockMap.forEach((k, v) -> v.lock());
+      return keyLockMap;
     }
-    return Collections.emptyList();
+    return Collections.emptyMap();
   }
 
-  public void unlockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records, List<ReentrantLock> locks) {
+  public void unlockKeys(Map<ByteArrayKey, ReentrantLock> keyLockMap) {
     if (lockManager != null) {
-      locks.forEach(lock -> lock.unlock());
-      records.forEach(r -> {
-        if (!r.getKey().isControlMessage()) {
-          lockManager.releaseLock(ByteArrayKey.wrap(r.getKey().getKey()));
-        }
+      keyLockMap.forEach((key, lock) -> {
```

Comment:
Nit, i don't think anything could go wrong, but if we release the locks in the reverse order of their acquisition, it might be easy to reason about in the future.